kubectl-cloudflow/verify/verify.go

```diff
@@ -0,0 +1,142 @@
+	}
+
+	// 1. checks images section
+	// 2. checks streamlets section
+	// 3. checks for streamlet descriptors
+	// 4. checks consistency between images section and image ids referred to in streamlets section
+	// 5. checks if the image referred to in a streamlet contains the streamlet descriptor in the label present in the image
+	// 6. checks connection problems
+	blueprint = blueprint.verify()
+	var errors string
+	for _, p := range blueprint.globalProblems {
+		errors = errors + p.ToMessage() + "\n"
+	}
+	if len(blueprint.globalProblems) == 0 {
+		return nil
+	}
+	return fmt.Errorf("%s", errors)
+}
+
+// this map is constructed entirely from blueprint
+func getImageRefsFromConfig(config *configuration.Config) map[string]domain.ImageReference {
+	// get the images from the blueprint
+	images := config.GetNode("blueprint.images").GetObject().Items()
+
+	var imageKeyVals = make(map[string]domain.ImageReference)
+
+	for imageKey, imageRef := range images {
+		ref, err := ParseImageReference(imageRef.GetString())
+		if err != nil {
+			util.LogAndExit(err.Error())
+		}
+		imageKeyVals[imageKey] = *ref
+	}
+	return imageKeyVals
+}
+
+func getConnectionsFromBlueprintConfig(config *configuration.Config) []StreamletConnection {
+	connectionMap := config.GetNode("blueprint.connections").GetObject().Items()
+	var conns []StreamletConnection
+
+	for fromStreamlet, rest := range connectionMap {
+		outs := rest.GetObject().Items()
+		for fromPort, ins := range outs {
+			for _, in := range ins.GetArray() {
+			  conns = append(conns, StreamletConnection{from: fmt.Sprintf("%s.%s", fromStreamlet, fromPort), to: in.GetString(), metadata: config})
+			}
+		}
+	}
+	return conns
+}
+
+func getConnectionsFromBlueprintConf(config *configuration.Config) []StreamletConnection {
+	connectionMap := config.GetNode("blueprint.connections").GetObject().Items()
+	var conns []StreamletConnection
+	for from, tos := range connectionMap {
+		for _, to := range tos.GetArray() {
+			conns = append(conns, StreamletConnection{from: from, to: to.GetString(), metadata: config})
+		}
+	}
+	return conns
+}
+
+func getStreamletRefsFromBlueprintConfig(config *configuration.Config) []StreamletRef {
+	streamlets := config.GetNode("blueprint.streamlets").GetObject().Items()
+	var streamletRefs []StreamletRef
+	for name, classWithImage := range streamlets {
+		arr := strings.Split(classWithImage.GetString(), "/")
+		streamletRef := StreamletRef{name: name, className: arr[1], imageId: &arr[0], metadata: config}
+		streamletRefs = append(streamletRefs, streamletRef)
+	}
+	return streamletRefs
+}
+
+func getStreamletDescriptorsFromImageRefs(imageRefs map[string]domain.ImageReference) map[string][]StreamletDescriptor {
+	apiversion, apierr := exec.Command("docker", "version", "--format", "'{{.Server.APIVersion}}'").Output()
+	if apierr != nil {
+		util.LogAndExit("Could not get docker API version, is the docker daemon running? API error: %s", apierr.Error())
```

Comment:
please return error

Reply:
no need in my opinion but since @debasishg has written this part let's discuss.

Reply:
My point is that testing these kind of functions can be useful too and it is easier to do (and just more appropriate IMHO) that if they return errors. It's not really about who wrote what.. anything that is found is valuable?

Reply:
@RayRoestenburg who ever wrote something has a better opinion so I am asking his ;)